File: mobile/lib/data/repositories/in_memory_course_repository.dart
import '../../domain/repositories/i_course_repository.dart';
import '../../domain/entities/course.dart';
import '../../domain/entities/skill.dart';
import '../../domain/entities/question.dart';
import '../datasources/courses_data.dart';
import '../datasources/skills_data.dart';
import '../datasources/questions_data.dart';

class InMemoryCourseRepository implements ICourseRepository {
  @override
  Future<List<Course>> getAllCourses() async {
    return List.unmodifiable(seedCourses);
  }

  @override
  Future<Course?> getCourseById(String courseId) async {
    return seedCourses.where((c) => c.id == courseId).firstOrNull;
  }

  @override
  Future<List<Skill>> getSkillsByCourseId(String courseId) async {
    return seedSkills.where((s) => s.courseId == courseId).toList();
  }

  @override
  Future<List<Question>> getPlacementQuestions(String courseId) async {
    return List.unmodifiable(placementQuestionsData[courseId] ?? []);
  }
}

File: src/data/repositories/InMemoryCourseRepository.ts
import { ICourseRepository } from '@/domain/repositories/ICourseRepository';
import { Course } from '@/domain/entities/Course';
import { Skill } from '@/domain/entities/Skill';
import { Question } from '@/domain/entities/Question';
import { SEED_COURSES } from '../datasources/courses';
import { SEED_SKILLS } from '../datasources/skills';
import { PLACEMENT_QUESTIONS } from '../datasources/questions';

export class InMemoryCourseRepository implements ICourseRepository {
  async getAllCourses(): Promise<Course[]> {
    return [...SEED_COURSES];
  }

  async getCourseById(courseId: string): Promise<Course | null> {
    const found = SEED_COURSES.find((c) => c.id === courseId);
    return found ? { ...found } : null;
  }

  async getSkillsByCourseId(courseId: string): Promise<Skill[]> {
    return SEED_SKILLS.filter((s) => s.courseId === courseId);
  }

  async getPlacementQuestions(courseId: string): Promise<Question[]> {
    return PLACEMENT_QUESTIONS[courseId] ? [...PLACEMENT_QUESTIONS[courseId]] : [];
  }
}

export const inMemoryCourseRepository = new InMemoryCourseRepository();
